server/server/src/billing/billing_service.rs

```diff
@@ -1,351 +1,358 @@
-                id: customer_id.to_string(),
-                last_4,
-                created_at: payment_method_resp.created as u64,
-            });
+            sub_profile.billing_platform = None;
 
-            (customer_id, payment_method_resp.id.to_string())
+            info!("Successfully canceled stripe subscription. public_key: {:?}", context.public_key);
         }
-        PaymentMethod::OldCard => {
-            info!("Using an old card stored on redis for public_key: {}", fmt_public_key);
-
-            let payment_method = user_info
-                .payment_methods
-                .iter()
-                .max_by_key(|info| info.created_at)
-                .ok_or(ClientError(SwitchAccountTierError::OldCardDoesNotExist))?;
-
-            match &user_info.customer_id {
-                Some(customer_id) => (customer_id.parse()?, payment_method.id.clone()),
-                None => return Err(internal!("StripeUserInfo is in an inconsistent state: has payment method but no customer id: {:?}", user_info))
-            }
-        }
-    };
-
-    info!("Successfully retrieved card for public_key: {}", fmt_public_key);
-
-    let subscription_resp =
-        stripe_client::create_subscription(server_state, customer_id.clone(), &payment_method_id)
-            .await?;
+    }
 
-    info!(
-        "Successfully create subscription: {}, public_key: {}",
-        subscription_resp.id, fmt_public_key
-    );
+    release_subscription_profile(&context.public_key, &mut con, &mut sub_profile).await?;
 
-    user_info.subscriptions.push(StripeSubscriptionInfo {
-        id: subscription_resp.id.to_string(),
-        period_end: subscription_resp.current_period_end as u64,
-        is_active: true,
-    });
-
-    Ok(PREMIUM_TIER_USAGE_SIZE)
+    Ok(CancelSubscriptionResponse {})
 }
 
-pub async fn get_credit_card(
-    context: RequestContext<'_, GetCreditCardRequest>,
-) -> Result<GetCreditCardResponse, ServerError<GetCreditCardError>> {
-    let mut con = context.server_state.index_db_pool.get().await?;
-
-    info!("Getting credit card for {}", keys::stringify_public_key(&context.public_key));
-
-    let user_info: StripeUserInfo = con
-        .maybe_json_get(stripe_user_info(&context.public_key))
-        .await?
-        .ok_or(ClientError(GetCreditCardError::NotAStripeCustomer))?;
-
-    let payment_method = user_info
-        .payment_methods
-        .iter()
-        .max_by_key(|info| info.created_at)
-        .ok_or(ClientError(GetCreditCardError::NotAStripeCustomer))?;
+async fn save_subscription_profile<
+    T: Debug,
+    F: Fn(&mut SubscriptionProfile) -> Result<(), ServerError<T>>,
+>(
+    public_key: &PublicKey, con: &mut deadpool_redis::Connection,
+    millis_between_payment_flows: u64, millis_between_lock_attempts: Duration,
+    update_subscription_profile: F,
+) -> Result<(), ServerError<T>> {
+    loop {
+        match lock_subscription_profile(public_key, con, millis_between_payment_flows).await {
```

Comment:
If you want to keep spinning for this lock, you should add a sleep. I think this function should be clear in it's name that it's looping trying to talk to the network and sleeping in the meantime. Without getting carried away in complexity, you also know how long you would have to wait. Also if this happens, you should def log the old value as it's being overwritten).

Reply:
I do sleep in this loop. If I can't get the billing lock, I sleep for a certain amount of time defined in the `BillingConfig`. If I do get the lock successfully, I will make my changes and stop there.

Reply:
Also, there are some awaits scattered throughout this function and lock_subscription_profile. So it shouldn't necessarily be blocking, right? In most cases, this function will pass off the first try anyway.